contracts/minter-suite/MinterHolderV0.sol

```diff
@@ -0,0 +1,494 @@
+     * prove right to purchase.
+     * @param _ownedNftTokenId ERC-721 NFT token ID owned by msg.sender being used
+     * to prove right to purchase.
+     * @return tokenId Token ID of minted token
+     */
+    function purchase(
+        uint256 _projectId,
+        address _ownedNftAddress,
+        uint256 _ownedNftTokenId
+    ) external payable returns (uint256 tokenId) {
+        tokenId = purchaseTo(
+            msg.sender,
+            _projectId,
+            _ownedNftAddress,
+            _ownedNftTokenId
+        );
+        return tokenId;
+    }
+
+    /**
+     * @notice Purchases a token from project `_projectId` and sets
+     * the token's owner to `_to`.
+     * @param _to Address to be the new token's owner.
+     * @param _projectId Project ID to mint a token on.
+     * @param _ownedNftAddress ERC-721 NFT address owned by msg.sender being used to
+     * claim right to purchase.
+     * @param _ownedNftTokenId ERC-721 NFT token ID owned by msg.sender being used
+     * to claim right to purchase. Must have not yet been used to claim a mint
+     * on project `_projectId`.
+     * @return tokenId Token ID of minted token
+     */
+    function purchaseTo(
+        address _to,
+        uint256 _projectId,
+        address _ownedNftAddress,
+        uint256 _ownedNftTokenId
+    ) public payable nonReentrant returns (uint256 tokenId) {
+        // CHECKS
+        require(
+            !projectMaxHasBeenInvoked[_projectId],
+            "Maximum number of invocations reached"
+        );
+
+        require(
+            msg.value >= projectIdToPricePerTokenInWei[_projectId],
+            "Must send minimum value to mint!"
+        );
+
+        // require artist to have configured price of token on this minter
+        require(
+            projectIdToPriceIsConfigured[_projectId],
+            "Price not configured"
+        );
+
+        // require token used to claim to be in set of allowlisted NFTs
+        require(
+            isAllowlistedNFT(_projectId, _ownedNftAddress, _ownedNftTokenId),
+            "Only allowlisted NFTs"
+        );
+
+        // EFFECTS
+        tokenId = minterFilter.mint(_to, _projectId, msg.sender);
+        // what if projectMaxInvocations[_projectId] is 0 (default value)?
+        // that is intended, so that by default the minter allows infinite transactions,
+        // allowing the artblocks contract to stop minting
+        // uint256 tokenInvocation = tokenId % ONE_MILLION;
+        if (
+            projectMaxInvocations[_projectId] > 0 &&
+            tokenId % ONE_MILLION == projectMaxInvocations[_projectId] - 1
+        ) {
+            projectMaxHasBeenInvoked[_projectId] = true;
+        }
+
+        // INTERACTIONS
+        // require sender to own NFT used to redeem
+        require(
+            IERC721(_ownedNftAddress).ownerOf(_ownedNftTokenId) == msg.sender,
+            "Only owner of NFT"
+        );
```

Comment:
I think I may be overlooking something simple here, but could you clarify why is in the interactions phase and not a check?

Reply:
@jakerockland I put this in interactions because we are calling `ownerOf` on a (allowlisted) token contract. Since we plan to limit this to Art Blocks+PBAB tokens, I can see why it could be classified as "checks" (since we should control the token contracts in this case).

Realistically, I don't think this check will fail very often, so I'm fine putting this anywhere. We also have nonReentrant, so there isn't a security risk placing this anywhere.

Reply:
Roger that–makes sense to me; I don't actually have a strong opinion here in either direction (within the known ecosystem of our contracts it is a check, but if expanded to things outside the ecosystem I def understand why we'd want to flag as an interaction). 

Happy to defer to your call on where to have this assertion live structurally (as CHECK vs. INTERACTIOn) but can we just elaborate on the inline documentation here to clarify with a comment the nuance we're capturing in our convo?

Reply:
I went ahead and left it as a interaction since I expect this check to very rarely cause a failure (meaning we won't be wasting gas), and added an @dev comment in the file 😄